src/backend/distributed/planner/distributed_planner.c

```diff
@@ -147,6 +148,112 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 }
 
 
+/*
+ * NeedsDistributedPlanning returns true if the Citus extension is loaded and
+ * the query contains a distributed table.
+ */
+bool
+NeedsDistributedPlanning(Query *query)
+{
+	CmdType commandType = query->commandType;
+	if (commandType != CMD_SELECT && commandType != CMD_INSERT &&
+		commandType != CMD_UPDATE && commandType != CMD_DELETE)
+	{
+		return false;
+	}
+
+	if (!CitusHasBeenLoaded())
+	{
+		return false;
+	}
+
+	/*
+	 * We can handle INSERT INTO distributed_table SELECT ... even if the SELECT
+	 * part references local tables, so skip the remaining checks.
+	 */
+	if (InsertSelectIntoDistributedTable(query))
+	{
+		return true;
+	}
+
+	if (!NeedsDistributedPlanningWalker((Node *) query, NULL))
+	{
+		return false;
+	}
+
+	if (InsertSelectIntoLocalTable(query))
+	{
+		ereport(ERROR, (errmsg("cannot INSERT rows from a distributed query into a "
+							   "local table")));
+	}
+
+	return true;
+}
+
+
+/*
+ * NeedsDistributedPlanningWalker checks if the query contains any distributed
+ * tables. Additionally, it errors out if there is a join between a local and
+ * a distributed table.
+ */
+static bool
+NeedsDistributedPlanningWalker(Node *node, void *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
+		ListCell *rangeTableCell = NULL;
+		bool hasLocalRelation = false;
+		bool hasDistributedRelation = false;
+
+		foreach(rangeTableCell, query->rtable)
+		{
+			RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
+			Oid relationId = InvalidOid;
+
+			if (rangeTableEntry->rtekind != RTE_RELATION ||
```

Comment:
My general aim is that we no longer prevent queries with local tables from reaching the distributed planner. I'd also like to get rid of the check for joins between local and distributed table here, but it simplifies a few things for now. 

> Whenever we find a local relation, we call ExtractRangeTableRelationWalker on that query to make sure that there are no other distributed tables in that query tree. If there is we error out. If there is not, we return false. Because, our end goal is to allow queries with only local tables, right?

It's not a given that a distributed table appearing in the query tree means that we cannot execute the query. It might be a subquery that we can (should?) recursively plan.

The only thing we can say conclusively at this point is that direct joins between local and distributed tables are unsupported. Everything else is a decision that should be made within the planner.

Reply:
> but it simplifies a few things for now.

On the other hand, makes the function slightly harder to follow. We sometimes allow local tables along with distributed tables and sometimes not. If the function were to return true as soon as it finds distributed table and had a comment for non-distributed tables (i.e., `is a decision that should be made within the planner.`), it might be simpler.

Also, what are those things that this check simplifies? Is that we'd need to move this check to multiple planners for this more useful error message compared to  `relation %s is not distributed`?

Reply:
I simplified the function and moved the INSERT ... SELECT error out as well. It's now a pristine function that just looks for distributed tables.

Reply:
As a side-effect, `REFRESH MATERIALIZED VIEW` is now supported, because the only thing blocking that was an extra range table entry (with rtekind `RTE_RELATION`). Added some regression tests.